src/tritonserver.cc

```diff
@@ -2157,6 +2157,112 @@ TRITONSERVER_ServerStop(TRITONSERVER_Server* server)
   return nullptr;  // Success
 }
 
+TRITONSERVER_DECLSPEC TRITONSERVER_Error*
+TRITONSERVER_ServerRegisterModelRepository(
+    TRITONSERVER_Server* server, const char* repository_path,
+    const TRITONSERVER_Parameter** name_mapping, const uint32_t mapping_count)
+{
+  tc::InferenceServer* lserver = reinterpret_cast<tc::InferenceServer*>(server);
+  if (lserver->GetModelControlMode() != tc::ModelControlMode::MODE_EXPLICIT) {
+    std::string mode;
+    switch (lserver->GetModelControlMode()) {
+      case tc::ModelControlMode::MODE_NONE:
+        mode = "NONE";
+        break;
+      case tc::ModelControlMode::MODE_POLL:
+        mode = "POLL";
+        break;
+      default:
+        mode = "UNKNOWN";
+        break;
+    }
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_UNSUPPORTED,
+        (std::string("Register API is unsupported in ") + mode +
+         " model control mode")
+            .c_str());
+  }
+  if ((name_mapping == nullptr) && (mapping_count != 0)) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        "model mappings are not provided while mapping count is non-zero");
+  }
+  bool is_directory = false;
+  auto status = triton::core::IsDirectory(repository_path, &is_directory);
+  if (!status.IsOk() || !is_directory) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        (std::string("failed to register '") + repository_path +
+         "', repository not found")
+            .c_str());
+  }
+
+  RETURN_IF_STATUS_ERROR(lserver->AddModelRepositoryPath(repository_path));
+  std::unordered_map<std::string, std::string> model_mapping;
```

Comment:
Why not just make one function `InferenceServer::AddModelRepositoryPath(const std::string& repository, std::unordered_map<std::string, std::string> model_mapping);`, then the implementation here is just translating Triton objects to C++ objects and propagating it downwards. Where the validations will be done in model manager.

In this way, you don't need to worry about unregistering the repository if error is returned

Reply:
Done.